### PR TITLE
Use dynamic storage for zerowidth characters

### PR DESCRIPTION
**Pull request by chrisduerr**
The zerowidth characters were conventionally stored in a `[char; 5]`.
This creates problems both by limiting the maximum number of zerowidth
characters and by increasing the cell size beyond what is necessary even
when no zerowidth characters are used.

Instead of storing zerowidth characters as a slice, a new `CellExtra`
struct is introduced which can store arbitrary optional cell data that
is rarely required. Since this is stored behind an optional pointer
(`Option<Box<CellExtra>>`), the initialization and dropping in the case
of no extra data are extremely cheap and the size penalty to cells
without this extra data is limited to 8 instead of 20 bytes.

The most noticible difference with this PR should be a reduction in
memory size of up to at least 30% (1.06G -> 733M, 100k scrollback, 72
lines, 280 columns).

### CODE REVIEW

**Comment by chrisduerr** [comment]
I'm not entirely sure if there are any large changes left for this PR, but I think it's at a point where a performance validation from multiple parties and on multiple operating systems makes sense.

So I'd love some feedback on the performance and the approach taken here (@kchibisov / @nixpulvis).

If someone is willing to test this with `vtebench`, I'd like the results with the following patch applied. The purpose of the patch is to ignore the performance difference from initial grid initialization, since that shouldn't bee too important and makes a rather big difference.

<details>

```diff
diff --git a/benchmarks/scrolling/setup b/benchmarks/scrolling/setup
new file mode 100755
index 0000000..f968ff1
--- /dev/null
+++ b/benchmarks/scrolling/setup
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+for _ in {0..100000}; do
+    echo "y"
+done
diff --git a/benchmarks/scrolling_fullscreen/setup b/benchmarks/scrolling_fullscreen/setup
new file mode 100755
index 0000000..f968ff1
--- /dev/null
+++ b/benchmarks/scrolling_fullscreen/setup
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+for _ in {0..100000}; do
+    echo "y"
+done
diff --git a/benchmarks/scrolling_top_region/benchmark b/benchmarks/scrolling_top_region/benchmark
deleted file mode 120000
index cfe87e7..0000000
--- a/benchmarks/scrolling_top_region/benchmark
+++ /dev/null
@@ -1 +0,0 @@
-../scrolling/benchmark
\ No newline at end of file
diff --git a/benchmarks/scrolling_top_region/setup b/benchmarks/scrolling_top_region/setup
deleted file mode 100755
index 11ecc0a..0000000
--- a/benchmarks/scrolling_top_region/setup
+++ /dev/null
@@ -1,3 +0,0 @@
-#!/bin/sh
-
-printf "\e[?1049h\e[2;$(tput lines)r"
diff --git a/benchmarks/scrolling_top_small_region/benchmark b/benchmarks/scrolling_top_small_region/benchmark
deleted file mode 120000
index cfe87e7..0000000
--- a/benchmarks/scrolling_top_small_region/benchmark
+++ /dev/null
@@ -1 +0,0 @@
-../scrolling/benchmark
\ No newline at end of file
diff --git a/benchmarks/scrolling_top_small_region/setup b/benchmarks/scrolling_top_small_region/setup
deleted file mode 100755
index 8bb1b50..0000000
--- a/benchmarks/scrolling_top_small_region/setup
+++ /dev/null
@@ -1,5 +0,0 @@
-#!/bin/sh
-
-lines=$(tput lines)
-
-printf "\e[?1049h\e[$((lines / 2));${lines}r"
```

</details>

Tests still fail, though that's just because the zerowidth test needs to be re-recorded. I'd propose doing that once everything else is mostly signed off. Otherwise that would need to be done multiple times which would be rather inefficient.

**Review by kchibisov — CHANGES_REQUESTED** [review]
We'd likely need a CHANGELOG entry, since it changes stuff...

File: alacritty_terminal/src/grid/row.rs
Comment (kchibisov) [inline alacritty_terminal/src/grid/row.rs:46]: Is it really necessary to do that and not use `resize`?

File: alacritty_terminal/src/grid/mod.rs
Comment (chrisduerr) [reply alacritty_terminal/src/grid/mod.rs:219]:
```rust
    pub fn scroll_down<I>(&mut self, region: &Range<Line>, positions: Line, template: I)
    where
        I: ResetDiscriminant<Color> + Into<T> + Copy,
```

```rust
    pub fn scroll_up<I>(&mut self, region: &Range<Line>, positions: Line, template: I)
    where
        I: ResetDiscriminant<Color> + Into<T> + Clone,
```

I don't see how they could be any more similar?

File: alacritty_terminal/src/grid/row.rs
Comment (chrisduerr) [reply alacritty_terminal/src/grid/row.rs:42]: This should always be guaranteed, due to our minimum grid size. I do not want to waste performance on an assertion that will never be hit anyways.

Though if it ever gets past this, it will crash terribly.

File: alacritty_terminal/src/term/cell.rs
Comment (chrisduerr) [reply alacritty_terminal/src/term/cell.rs:50]: I feel like we can leave that up to applications? Like if they really want to put 60 in there, that's fine?

If we find out that a lot of applications might be misbehaving, we can change that in the future, but I don't see a reason to immediately restrict it.

File: alacritty_terminal/src/grid/row.rs
Comment (chrisduerr) [reply alacritty_terminal/src/grid/row.rs:46]: Yes.

Comment (chrisduerr) [reply alacritty_terminal/src/grid/row.rs:46]: Unless you'd prefer slower code of course. Then we could use `resize` instead.

**Comment by nixpulvis** [comment]
Here's the output of a couple runs of `vtebench` on my machine on both 8924b757 and f33a44e1.


summary

**Comment by kchibisov** [comment]
My output of vtebench on my machines on both 8924b75 and f33a44e.

linux-desktop-4700

linux-desktop-3600

linux-laptop

macos-laptop


The only suspicious thing here is that without turbo my desktop had slower scrolling, but with it it was faster...

As for macOS and a laptop(4750u) I can't really fix frequency on both, so they just use `+` to indicate that, the same applies to my desktop under turbo(here I'm just lazy to do that properly for turbo, since requires reboot).

**Comment by chrisduerr** [comment]
> The only suspicious thing here is that without turbo my desktop had slower scrolling, but with it it was faster...

First of all that's some very detailed collection of results, thanks for going so in depth with the different frequencies. Have you only run every benchmark once, or did you verify that this is repeatable?

macOS and Linux laptop both seem great with performance I wouldn't mind merging, but the scrolling performance without turbo seems to be down by around 15%, which is more than I'd be willing to take.

@nixpulvis It seems like you forgot to apply my patch? That probably means these results don't really say much. Though I like the idea of running the short benchmark a couple of times rather than just running it for a longer time as a demonstration of consistency.

**Comment by kchibisov** [comment]
>macOS and Linux laptop both seem great with performance I wouldn't mind merging, but the scrolling performance without turbo seems to be down by around 15%, which is more than I'd be willing to take.

I may rerun, but IIRC it was consistent, I'm not sure that you can really do anything, assuming that you scrolling is faster with turbo...

**Review by kchibisov — CHANGES_REQUESTED** [review]
I've tested perf more and it seems fine to me.

The PR is till missing CHANGELOG entries (for stuff like memory usage and more zerowidth per cells).

I also left some more minor comments.

**Comment by nixpulvis** [comment]
OK, I've rerun the benchmarks on both my i5 desktop and my i7U laptop with the vtebench patch as mentioned. When and if the macOS benches are fixed I'll report those for my 2012 MacBook Pro as well.

summary

summary

File: alacritty_terminal/src/grid/storage.rs
Comment (cole-h) [reply alacritty_terminal/src/grid/storage.rs:69]: It's correct.

File: alacritty_terminal/src/grid/row.rs
Comment (chrisduerr) [reply alacritty_terminal/src/grid/row.rs:113]: Basically that's the idea, yes.

File: alacritty_terminal/src/grid/storage.rs
Comment (chrisduerr) [reply alacritty_terminal/src/grid/storage.rs:69]: Yeah, it's basically a completely separate sentence but it is still related to the first one, that's why I've put the semicolon there. With a comma I'd have to put a `since` or something in between for it to read nicely imo.

File: alacritty_terminal/src/term/mod.rs
Comment (chrisduerr) [reply alacritty_terminal/src/term/mod.rs:857]: I've wanted to remove this myself actually, completely agree.

File: alacritty_terminal/src/term/mod.rs
Comment (chrisduerr) [reply alacritty_terminal/src/term/mod.rs:258]: I can add it back, no problem. I just removed it to make it easier for me to understand what is going on and where exactly things are cloned around.

**Comment by chrisduerr** [comment]
@nixpulvis Seems like it's basically slower in every way for you? Especially the `scrolling_fullscreen` benchmark seems very worrying.

**Comment by chrisduerr** [comment]
I've decided to fix the zerowidth tests, for reference, this is all I had to do to rewrite them:

```bash
for file in alacritty_terminal/tests/ref/*/grid.json; do
    sed -i 's/,"extra":\[[^]]*]//g' "$file";
done
```

The only file that was re-recorded is the `zerowidth` one. With all others just removing the extra data was sufficient and should be reproducible when copying the latest master and running that `sed` loop.

**Comment by chrisduerr** [comment]
We could reduce the average execution time of the grid reset during the scrolling benchmark by about 2ms (~1%) by passing a template `Color` as a parameter instead of using `T::default()`. In theory the same would apply to the grid resize, though I have no good benchmarks for that.

I personally do not think this is worth it, but here are the results of my test:

tmp

File: alacritty_terminal/src/grid/resize.rs
Comment (chrisduerr) [inline alacritty_terminal/src/grid/resize.rs:17]: This is the biggest drawback of the new approach I'd say. Since we just access the cursor template inside the grid automatically without a parameter this requires overwriting the template cell and then resetting it again later.

Generally this isn't that bad though I'd say and it makes sense since we want to switch the template context during reset, so it correctly represents what we're trying to do.

There are two alternatives, which both were significantly worse in my tests:

One could still take a `&T` as template everywhere. But since we pass these in from `term/mod.rs` we'd have to pass `&self.grid.cursor.template`, which obviously creates mutability problems. So we'd have to clone the cursor template just to pass a reference to it. Obviously not very nice.

One thing I have tested was passing a `template_override: Option<&T>` in as an argument. This allows replacing the template on demand when necessary with `Some(&T::default())`, but pass `None` whenever we want to just use the grid's template. Then we can just `template_override.unwrap_or(&self.cursor.template)` and everything's good to go. 

I personally think that approach would have been okay, but the additional parameters make it a bit more annoying to work with and somewhat strange. Ultimately the reason why I had to discard it was due to performance limitations. Since most of the time we do not have a template override, it's just wasteful to check. And that made a big performance difference.

**Review by chrisduerr — COMMENTED** [review]
I've reworked everything completely to take a `&T` instead of the extremely noisy and complicated `Into<T>` for resetting everything. Wherever possible I've also used `T::default()` instead of any input template to reduce the number of parameters when only an initialization to any arbitrary value is required.

In theory `Into<T>` is just 4 bytes when it is a `Color` which is smaller than the pointer size of 8 bytes on x86_64, but since we're talking about a 64 bit architecture it's obviously pretty good at passing around 8 byte pointers. I wasn't able to find any significant performance difference due to this size difference.

The grid now also doesn't take a template cell for its arguments anymore, instead it just uses the template cell stored on the grid itself. This obviously makes a lot of sense since that's what it is there for but comes with a slight caveat (see my other comment).

I think this should provide a much, much nicer code and based on my testing it's identical to the previous non_copy_cells implementation (not identical to master). You're all welcome to benchmark this again; I will attach my own once they're done.

@kchibisov What do you think about the new code vs the previous version?

**Comment by chrisduerr** [comment]
These are the latest performance results, the latest version is named `non_copy_madness` since that was my working branch name:

tmp

**Comment by kchibisov** [comment]
I've rerun benchmarks and here are results.







As you can see, it kind of matches to what you're getting.

**Comment by chrisduerr** [comment]
tmp

For reference, these are my latest performance results with a little idea for a possible future performance enhancement by inlining the `width` function into the `input` function (or alternatively adding `#[inline(never)]` to the `input` function).

Though while most benchmarks like light_cells, dense_cells, and scrolling_fullscreen seem to at least slightly profit, there are other benchmarks like cursor_motion which do not seem to benefit from it.

File: alacritty_terminal/src/grid/storage.rs
Comment (kchibisov) [inline alacritty_terminal/src/grid/storage.rs:132]: That seems redundant.

File: alacritty_terminal/src/term/cell.rs
Comment (kchibisov) [inline alacritty_terminal/src/term/cell.rs:98]: Why not  always do `self.extra = None`, should be faster than useless branch here?

**Review by kchibisov — CHANGES_REQUESTED** [review]

File: alacritty_terminal/src/grid/mod.rs
Comment (chrisduerr) [reply alacritty_terminal/src/grid/mod.rs:264]: The goal is just to be uniform. It's either all in the first, or all in the second definition, not mixed and matched.

Here `ResetDiscriminant` is so long, that putting everything in the `where` is the better option.

File: CHANGELOG.md
Comment (chrisduerr) [reply CHANGELOG.md:39]: Great point.

File: alacritty_terminal/src/grid/storage.rs
Comment (chrisduerr) [reply alacritty_terminal/src/grid/storage.rs:132]: Probable seems that way because it is.

File: alacritty_terminal/src/term/cell.rs
Comment (chrisduerr) [reply alacritty_terminal/src/term/cell.rs:98]: Because it's slower:

tmp

tmp

Comment (kchibisov) [reply alacritty_terminal/src/term/cell.rs:98]: That it's so strange, but whatever. Like `None = None` should be the same...

Comment (chrisduerr) [reply alacritty_terminal/src/term/cell.rs:98]: I'm not entirely sure how it is implemented, but it seems to do more than just check for emptiness itself.

**Review by kchibisov — APPROVED** [review]

**Comment by chrisduerr** [comment]
Since it does provide a good performance improvement without any performance disadvantages, I have decided to add the `#[inline(never)]` annotation to the input function. This is something we should likely monitor in the future, together with the inline annotation on `write_at_cursor`.

tmp

As you can see in this comparison, the `#[inline]` annotation is reliably the worst option. While no attribute would be preferred for maintenance overhead in the future, the significant hit it takes during scrolling does not seem to be worth it. I've run these benchmarks three times and they are reliable, so moving forward with `#[inline(never)]` seems like the best choice.

**Pull request merged**
